src/Compiler/ES/ES.idr

```diff
@@ -194,50 +194,64 @@ jsConstant (B32 i) = pure $ show i ++ "n"
 jsConstant (B64 i) = pure $ show i ++ "n"
 jsConstant ty = throw (InternalError $ "Unsuported constant " ++ show ty)
 
+-- Creates the definition of a binary arithmetic operation.
+-- Rounding / truncation behavior is determined from the
+-- `IntKind`.
+arithOp :  {auto c : Ref ESs ESSt}
+        -> Maybe IntKind
+        -> (op : String)
+        -> (x : String)
+        -> (y : String)
+        -> Core String
+arithOp (Just $ Signed $ P n)   op x y = boundedIntOp (n-1) op x y
+arithOp (Just $ Unsigned $ P n) op x y = boundedUIntOp n op x y
+arithOp _                       op x y = pure $ binOp op x y
+
+-- Same as `arithOp` but for bitwise operations that might
+-- go out of the valid range.
+bitOp :  {auto c : Ref ESs ESSt}
+      -> Maybe IntKind
+      -> (op : String)
+      -> (x : String)
+      -> (y : String)
+      -> Core String
+bitOp (Just $ Signed $ P n)   op x y = boundedIntBitOp (n-1) op x y
+bitOp (Just $ Unsigned $ P n) op x y = boundedUIntOp n op x y
+bitOp _                       op x y = pure $ binOp op x y
+
+castInt :  {auto c : Ref ESs ESSt}
+        -> Constant
+        -> Constant
+        -> String
+        -> Core String
+castInt from to x =
+  case (intKind from, intKind to) of
+       (Just _, Just $ Signed Unlimited) => pure x
+
+       (Just $ Signed m, Just $ Signed $ P n) =>
+         if P n >= m then pure x else boundedInt (n-1) x
+
+       -- Only if the precision of the target is greater
+       -- than the one of the source, there is no need to cast.
+       (Just $ Unsigned m, Just $ Signed $ P n) =>
+         if P n > m then pure x else boundedInt (n-1) x
+
+       (Just $ Signed _, Just $ Unsigned $ P n) => boundedUInt n x
+
+       (Just $ Unsigned m, Just $ Unsigned $ P n) =>
+         if P n >= m then pure x else boundedUInt n x
+
+       _ => jsCrashExp $ jsString $ "invalid cast: + " ++ show from ++ " + ' -> ' + " ++ show to
```

Comment:
Should this be a `throw`? The backends tend, too often, to happily generate garbage
rather than complaining that an error occurred and that makes debugging pretty painful.

Reply:
I'll look into this. Where do I find out more about `throw`?

Reply:
Should be fixed now.